iceoryx_utils/include/iceoryx_utils/cxx/sized_uninitialized_array.hpp

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2021 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ICEORYX_UTILS_CXX_SIZED_UNINITIALIZED_ARRAY_HPP
+#define ICEORYX_UTILS_CXX_SIZED_UNINITIALIZED_ARRAY_HPP
+
+#include "iceoryx_utils/cxx/sized_container.hpp"
+#include <cstdint>
+
+namespace iox
+{
+namespace cxx
+{
+template <typename T, uint64_t Capacity>
+class SizedUninitializedArray : public SizedContainer<T, Capacity>
```

Comment:
What was your intention when you created `SizedContainer`?
I do not fully understand your design and at the first glimpse it makes for me more sense to move the implementation of `SizedContainer` directly into the `SizedUninitializedArray`.

Additionally, I would overload `SizedUninitializedArray` for the zero use case. With this approach we would get rid of the `m_data` member completely! The trick with `element_t m_data[Capacity == 0 ? 1 : Capacity]` would be a good approach when we wanted to solve the problem directly in the vector but when we create an abstraction we should go for the specialization.

Something like this:

```cpp
template <typename T, uint64_t Capacity>
class SizedUninitializedArray;

template <typename T>
class SizedUninitializedArray<T, 0> {
  public:
     uint64_t capacity() const noexcept {
           return 0U;
     }
     // and data(), max_size()...
};

template <typename T, uint64_t Capacity>
class SizedUninitializedArray<T, Capacity> {
  public:
     uint64_t capacity() const noexcept {
           return Capacity;
     }
     // and data(), max_size()...
};
```

Reply:
the specialization is not necessary since an instance of that specialization would have the same size as the current implementation. We wouldn't gain anything but add code.

Reply:
@elBoberido argh, you are right ... I missed that we cut it to one byte.

Reply:
If we merge `SizedContainer` and `SizedUninitializedArray`, we would need to specialize this new templated class for Capacity 0. This would imply that the 4 methods of `SizedUninitializedArray` would be duplicated. The "trick" with m_data avoids this and does not increase memory consumption, since an empty struct also has size 1.
Do you prefer to merge `SizedContainer` into `SizedUninitializedArray`?

Reply:
@elBoberido but I would still move the `SizedContainer` impl into the `SizedUninitializedArray`.

Reply:
Like I already wrote in the `SizedContainer` class. Don`t inherit from it. It makes this building block less flexible.

Reply:
> @elBoberido but I would still move the `SizedContainer` impl into the `SizedUninitializedArray`.

@elfenpiff no, it doesn't belong into `SizedUninitializedArray`, neither should `SizedUninitializedArray` inherit from it. They are not related to each other. An array has a capacity, not a size to set at runtime. That's what a vector is or a storage used to build a vector but not an array.

Reply:
See https://github.com/eclipse-iceoryx/iceoryx/pull/491#discussion_r560183825.